File: resize/resize_test.go
package resize_test

import (
	"image"
	"image/color"
	"image/png"
	"os"
	"path/filepath"
	"testing"

	"github.com/arinji2/img-processor/convert"
	"github.com/arinji2/img-processor/resize"
	"golang.org/x/image/webp"
)

func createTestImage(path string) error {
	// Create a simple test image
	img := image.NewRGBA(image.Rect(0, 0, 800, 800))
	for x := range make([]struct{}, 800) {
		for y := range make([]struct{}, 600) {
			img.Set(x, y, color.RGBA{uint8(x % 256), uint8(y % 256), 100, 255})
		}
	}

	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()
	return png.Encode(file, img)
}

func TestResizeWebP(t *testing.T) {
	tempDir := t.TempDir()

	// First create a PNG image
	pngPath := filepath.Join(tempDir, "test.png")
	if err := createTestImage(pngPath); err != nil {
		t.Fatalf("Failed to create test image: %v", err)
	}

	// Convert PNG to WebP first
	webpPath := filepath.Join(tempDir, "test.webp")
	if err := convert.ConvertPNGToWebP(pngPath, webpPath); err != nil {
		t.Fatalf("Failed to convert PNG to WebP: %v", err)
	}

	// Now resize the WebP image
	sizes := map[string]uint{
		"sm": 480,
		"md": 768,
		"lg": 1024,
		"xl": 1280,
	}

	if err := resize.ResizeWebP("test", webpPath, sizes); err != nil {
		t.Fatalf("ResizeWebP failed: %v", err)
	}

	// Verify the output files exist and have the correct dimensions
	for sizeName, expectedWidth := range sizes {
		outputPath := filepath.Join(tempDir, "test_"+sizeName+".webp")

		// Check if file exists
		if _, err := os.Stat(outputPath); os.IsNotExist(err) {
			t.Errorf("Expected output file %s not found", outputPath)
			continue
		}

		// Open the file and verify dimensions
		file, err := os.Open(outputPath)
		if err != nil {
			t.Errorf("Failed to open output file %s: %v", outputPath, err)
			continue
		}

		// Decode the WebP image
		img, err := webp.Decode(file)
		file.Close()
		if err != nil {
			t.Errorf("Failed to decode WebP file %s: %v", outputPath, err)
			continue
		}

		// Check dimensions
		bounds := img.Bounds()
		actualWidth := bounds.Max.X - bounds.Min.X

		// Allow for a small margin of error due to rounding
		if uint(actualWidth) != expectedWidth {
			t.Errorf("Image %s has incorrect width: expected %d, got %d",
				outputPath, expectedWidth, actualWidth)
		}
	}
}

File: __tests__/resize.test.ts
import fs from "fs-extra";
import path from "path";
import sharp from "sharp";
import { resizeWebP } from "../src/resize";
import { createTestImage } from "./helpers/create-image";

describe("resizeWebP", () => {
  const tempDir = path.join(__dirname, "temp");
  const inputPath = path.join(tempDir, "test.webp");
  const sizes = {
    sm: 480,
    md: 768,
    lg: 1024,
    xl: 1280,
  };

  beforeAll(async () => {
    await fs.ensureDir(tempDir);
    await createTestImage(path.join(tempDir, "test.png"), "png");
    await sharp(path.join(tempDir, "test.png"))
      .webp({ quality: 95 })
      .toFile(inputPath);
  });

  afterAll(async () => {
    await fs.remove(tempDir);
  });

  it("should resize the image to all specified sizes", async () => {
    await resizeWebP("test", inputPath, sizes, tempDir);

    for (const [sizeName, expectedWidth] of Object.entries(sizes)) {
      const outputPath = path.join(tempDir, `test_${sizeName}.webp`);

      expect(fs.existsSync(outputPath)).toBe(true);

      const image = sharp(outputPath);
      const metadata = await image.metadata();

      expect(metadata.width).toBe(expectedWidth);
      expect(metadata.format).toBe("webp");
    }
  });
});
